### PR TITLE
Distributed reaper with PostgreSQL backend

### PR DESCRIPTION
**Pull request by Saleil-Bhat**
This PR implements IDistributedStorage for PostgreSQL; it enables the use of multiple, distributed reaper instances with a Postgres backend.

### CODE REVIEW

**Comment by michaelsembwever** [comment]
@Saleil-Bhat integration tests are flakey. try rebuilding, and if it still fails look into the logs for errors and see if it makes sense.

**Review by adejanovski — COMMENTED** [review]
Hi @Saleil-Bhat,

thanks for posting that PR!
I have a minor comment about introducing a new storage type, which doesn't seem to be necessary.
Also, we would need to have an additional integration test for that feature, in the `Distributed reaper integration tests` section, but using Postgres as storage. Would you mind adding it?

File: src/server/src/main/java/io/cassandrareaper/ReaperApplication.java [hunk 1]
```diff
@@ -346,13 +348,29 @@ private IStorage initializeStorage(ReaperApplicationConfiguration config, Enviro
       storage = new MemoryStorage();
     } else if ("cassandra".equalsIgnoreCase(config.getStorageType())) {
       storage = new CassandraStorage(config, environment);
+    } else if ("multireaper-postgres".equalsIgnoreCase(config.getStorageType())) {
+      // create DBI instance
```
[the remark this answers is not in the document]
Comment (Saleil-Bhat) [reply src/server/src/main/java/io/cassandrareaper/ReaperApplication.java]: Agreed. It makes logical sense for this functionality to be in the Postgres storage type; I'll make that change.

File: src/server/src/main/java/io/cassandrareaper/ReaperApplication.java [hunk 2]
```diff
@@ -396,7 +414,12 @@ private void initDatabase(ReaperApplicationConfiguration config) throws ReaperEx
         dsfactory.getUser(),
         dsfactory.getPassword());
 
-    if ("database".equals(config.getStorageType())) {
+    String storageDir = config.getStorageType().toLowerCase();
+    if (storageDir.contains("postgres")) {
+      storageDir = "postgres";
```
[the remark this answers is not in the document]
Comment (Saleil-Bhat) [reply src/server/src/main/java/io/cassandrareaper/ReaperApplication.java]: Yep

**Comment by adejanovski** [comment]
Hi @Saleil-Bhat, 

could you update me with the state of this PR? Do you need assistance to move forward with it?

Thanks!

**Comment by adejanovski** [comment]
Hi @Saleil-Bhat,

@michaelsembwever and I are close to merging the sidecar mode PR.
We have a lot of fixes there for the integration tests which will be helpful for getting this PR merged.
Let's rebase on top of master once the sidecar gets merged and spend some time together to get your PR finally merged.
Sorry for all the delays!

**Comment by adejanovski** [comment]
Hi @Saleil-Bhat, 

We're ready to resume work on this PR. Could you rebase your branch on top of master (you'll have a few conflicts to deal with)?
We made a lot of changes to implement the sidecar but it should have limited impact on your work.
We've also stabilized integration tests A LOT, so we mostly have no more flakey tests.

**Comment by Saleil-Bhat** [comment]
Ok great. I'll have time to work on this next week. I'll rebase and push then.

**Comment by Saleil-Bhat** [comment]
I've done a first-pass of rebasing this PR on sidecar mode. It passes all the travis-ci tests, but it's still a work in progress. The remaining issues are: 
1. Implement get / put node operations in Postgres storage (I wasn't sure what exactly you were getting at there)
2. Optimize data model for node_metrics_v2 (I'm just dumping them all in one table)
3. takeLead(int ttl) interface is implemented differently with Postgres than it is with Cassandra (since Postgres does not have TTL).  Right now, takeLead(int ttl) doesn't indicate the expiration time of the row to be inserted; rather, it indicates how old an already existing leadership row has to be for the Reaper executing takeLead to clobber it. This results in the same behavior as the Cassandra implementation because the config for all Reapers is the same; but it would cause issues if, for example,  one Reaper inserted a row with an expected ttl of 10 minutes, but another Reaper tried to take leadership of that row with a ttl of 5 minutes. 

Let me know you guys have any input on these issues.

**Comment by adejanovski** [comment]
> 1. Implement get / put node operations in Postgres storage (I wasn't sure what exactly you were getting at there)

In sidecar mode, each Reaper instance can only communicate with the local node. To get a cluster wide view of operations like ongoing compactions or streaming sessions, we need to store them in the backend. These are the only types of operations we're storing for now.
The lifecycle of such data is very short and just as for `node_metrics_v2`, which gives a cluster wide view of select metrics, you'll need to purge outdated data in the Postgres table on a very regular basis.
If I was implementing this, I would add appropriate purge methods in the `IDistributedStorage` interface, implement them as a noop in `CassandraStorage` (this is dealt with using TTLs), implement them in PostgresStorage to get rid of outdated data (you can use the TTL we use in Cassandra to filter rows to purge), and call them in the [PurgeService](https://github.com/thelastpickle/cassandra-reaper/blob/master/src/server/src/main/java/io/cassandrareaper/service/PurgeService.java) class. 
Purge runs every hour.

> 2. Optimize data model for node_metrics_v2 (I'm just dumping them all in one table)

Sounds good to me, and you'll also need to purge those.

> 3. takeLead(int ttl) interface is implemented differently with Postgres than it is with Cassandra (since Postgres does not have TTL).  Right now, takeLead(int ttl) doesn't indicate the expiration time of the row to be inserted; rather, it indicates how old an already existing leadership row has to be for the Reaper executing takeLead to clobber it. This results in the same behavior as the Cassandra implementation because the config for all Reapers is the same; but it would cause issues if, for example,  one Reaper inserted a row with an expected ttl of 10 minutes, but another Reaper tried to take leadership of that row with a ttl of 5 minutes.

Such TTLs aren't configurable, they are hardcoded. I don't think it that's a problem then.

You'll also need to add a test feature for the sidecar with Postgres as backend, distributed tests with Postgres and Flapping reapers with Postgres. Let's just use a single version of Cassandra for this otherwise we'll have too many tests and that'll take ages to complete.
The latest 3.11.x sounds like a good choice.

Thanks!

**Comment by Saleil-Bhat** [comment]
Regarding node operations:
For the node_operations table, the data is partitioned into hour-long buckets, but the TTL is only 5 minutes. What is the reason for this? Are node operations expected to persist for an hour or only 5 minutes? 
Also, I noticed that the CassandraStorage implementation of listOperations uses LIMIT 1 and that storeOperations(cluster, host, type, time_bucket) will clobber the previous value if it still belongs in the same time bucket. Does this mean there is only one valid value for a given operation on a given node at a time? 

Regarding node_metrics_v2: 
What do the fields 'domain', 'type', 'scope', and 'attribute' mean?

**Comment by adejanovski** [comment]
> Regarding node operations:
> For the node_operations table, the data is partitioned into hour-long buckets, but the TTL is only 5 minutes. What is the reason for this? Are node operations expected to persist for an hour or only 5 minutes?

We're only keeping the last value for each partition as Reaper instances will update each of them every 10 seconds. The buckets could be shorter indeed, but it works as it is (buckets are easy to compute as is). We're storing the current state of operations, not trying to keep their histories.
Using hourly buckets along with TWCS is an optimization to get efficient purges since data expires quickly.

> Also, I noticed that the CassandraStorage implementation of listOperations uses LIMIT 1 and that storeOperations(cluster, host, type, time_bucket) will clobber the previous value if it still belongs in the same time bucket. Does this mean there is only one valid value for a given operation on a given node at a time?

Yes, we're only interested in the current state of each operation, history is irrelevant. For more context, this data is used in the node detailed view, in tabs "compactions" and "streaming", to display running operations on remote nodes.

> 
> Regarding node_metrics_v2:
> What do the fields 'domain', 'type', 'scope', and 'attribute' mean?

That's how JMX metrics are organized.
For example, the following metric : `org.apache.cassandra.metrics:type=ColumnFamily,keyspace=system_auth,scope=roles,name=LiveSSTableCount` 

has: 

* domain = `org.apache.cassandra.metrics`
* type = `ColumnFamily`
* scope = `roles`

And it has a single attribute called `Value` (you can check that in `jconsole`).

**Comment by Saleil-Bhat** [comment]
I've addressed the 3 points discussed above and the PR now passes all the Travis CI tests (including Sidecar/Distributed/Flapping with Postgresql backend). I have two more questions, however: 
1. my implementation of node_metrics_v2 requires Postgresql 9.5+, as it uses the ON CONFLICT feature. Is this acceptable? 
2. I originally implemented the purge of node_metrics_v1 in PostgresStorage heartbeat, but now that purge of node_metrics_v2 is conducted in the PurgeService, it might make more sense to move that to the PurgeService as well. Do you have a particular preference?

**Comment by adejanovski** [comment]
> I've addressed the 3 points discussed above and the PR now passes all the Travis CI tests (including Sidecar/Distributed/Flapping with Postgresql backend). I have two more questions, however:
> 
> 1. my implementation of node_metrics_v2 requires Postgresql 9.5+, as it uses the ON CONFLICT feature. Is this acceptable?

What is the level of effort to support previous versions of Postgresql? We need to consider this before breaking compatibility. We may make some people very sad once they realize they need to upgrade Postgresql to upgrade Reaper.

> 2. I originally implemented the purge of node_metrics_v1 in PostgresStorage heartbeat, but now that purge of node_metrics_v2 is conducted in the PurgeService, it might make more sense to move that to the PurgeService as well. Do you have a particular preference?

It would be good indeed to have everything in the `PurgeService` if it is purge related. I hope that's not too much effort.

Thanks for all the hard work on this feature!
Looking forward to finally merge it 🙂

**Comment by Saleil-Bhat** [comment]
> 
> What is the level of effort to support previous versions of Postgresql? We need to consider this before breaking compatibility. We may make some people very sad once they realize they need to upgrade Postgresql to upgrade Reaper.
> 
Perhaps not so much. I used a somewhat normalized data model; each metric references an entry in the source node table.  Since nodes can leave clusters, these entries have a last_updated column and the purge operation deletes old entries. This means every insert is an upsert (where the update sets last_updated to the current time) which must be resilient to concurrent upserts and deletes.  However, upon further reflection, using Postgres upsert may not be necessary; I might be able to get away with: 
`try update; if rows_updated = 0 then insert`
For concurrent upserts, both instances are inserting the same data, so we don't care if one of the inserts fails. For deletes, either the delete happens first (in which case the upsert is just an insert) or the update happens first (in which case the delete will not affect the row).  

Let me know what you think. 

>
> It would be good indeed to have everything in the PurgeService if it is purge related. I hope that's not too much effort.
>
This is pretty easy to do; I will make the change.

**Comment by Saleil-Bhat** [comment]
Per the comments above, I re-implemented node_metrics_v2 with Postgres without using the Postgres 9.5+ upsert feature.   
I'm having trouble getting all the integration tests to pass because of some Travis before_install weirdness, but it looks like the key relevant tests are all passing. 
If someone could take a look at this, that'd be great

**Comment by adejanovski** [comment]
Sure thing @Saleil-Bhat, and thanks for the update that will give wider postres compatibility 👍 
I'll check the PR, run some tests and then hopefully merge it.

**Comment by adejanovski** [comment]
Hi Saleil, it's a fantastic job!

I have two last requests and then we'll merge once the Travis tests all pass (Travis is really flakey these days):

- Could you only retain "Postgresql reaping Cassandra 3.11.3" in the "Integration tests" section? The others aren't useful as all the versions of Cassandra are already tested. I'll do the same with Memory and H2, as this is a massive waste of time and builds are already taking ages to complete.
- Could you squash your 32 commits into 1 before we merge, so we can have a clean history?

And that'll be all 🎉

**Comment by Saleil-Bhat** [comment]
Ok made the requested changes. Just waiting on a good Travis run then.

**Review by adejanovski — APPROVED** [review]

**Pull request merged**